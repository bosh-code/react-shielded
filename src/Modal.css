.overlay {
  display: flex;
  flex-direction: column;
  align-items: center;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 99999;
  overflow: auto;
  background-color: rgba(0, 0, 0, 0.5);
}

.modal {
  width: 310px;
  display: flex;
  flex-direction: column;
  position: relative;
}

.loading {
  position: absolute;
  height: 420px;
  width: 310px;
  display: flex;
  justify-content: center;
  align-items: center;
}

@keyframes spin {
  0% {
    transform: rotate(0);
  }
  100% {
    transform: rotate(1turn);
  }
}

.spinner {
  animation: spin 1.5s infinite linear;
}

.frame {
  display: flex;
  flex-direction: column;
}

.close {
  appearance: none;
  background: none;
  display: inline-block;
  border: none;
  margin: 0;
  padding: 0;
  cursor: pointer;
  align-self: center;
  margin-top: -1px;
}
