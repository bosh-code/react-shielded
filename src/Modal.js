import React, { useEffect, useState, useRef } from 'react';
import useKeypress from 'react-use-keypress';
import Close from './close.svg';
import styles from './Modal.css';
import Portal from './Portal';
import Spinner from './spinner.svg';
import useLockBodyScroll from './useLockBodyScroll';

const Modal = (props) => {
  const { onRequestClose } = props;

  const iframeRef = useRef(null);

  const [isLoading, setLoading] = useState(true);

  useEffect(() => {
    if (isLoading === false) {
      iframeRef.current.focus();
    }
  }, [isLoading]);

  useKeypress('Escape', onRequestClose);

  useLockBodyScroll();

  return (
    <Portal>
      <div aria-modal="true" className={styles.overlay} role="dialog">
        <div className={styles.modal}>
          {isLoading && (
            <div className={styles.loading}>
              <Spinner
                aria-label="Loading"
                className={styles.spinner}
                role="img"
              />
            </div>
          )}
          <div className={styles.frame} style={{ opacity: isLoading ? 0 : 1 }}>
            <iframe
              frameBorder="0"
              height="420"
              onError={() => onRequestClose()}
              onLoad={() => setLoading(false)}
              ref={iframeRef}
              sandbox="allow-forms allow-scripts allow-same-origin allow-popups"
              src="https://staticcdn.co.nz"
              width="310"
            />
            <button
              type="button"
              className={styles.close}
              onClick={() => onRequestClose()}
            >
              <Close aria-label="Close" role="img" />
            </button>
          </div>
        </div>
      </div>
    </Portal>
  );
};

export default Modal;
