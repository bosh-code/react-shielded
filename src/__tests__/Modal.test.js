import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Modal from '../Modal';

test('renders and matches snapshot', () => {
  const { getByRole } = render(<Modal onRequestClose={() => {}} />);
  expect(getByRole('dialog')).toMatchSnapshot();
});

test('iframe is focused on mount', () => {
  const { getByLabelText } = render(<Modal onRequestClose={() => {}} />);
  fireEvent.load(getByLabelText('The Shielded Site'));
  expect(getByLabelText('The Shielded Site')).toHaveFocus();
});

test('loading indicator is hidden when iframe has loaded', () => {
  const { getByLabelText, queryByLabelText } = render(
    <Modal onRequestClose={() => {}} />
  );
  fireEvent.load(getByLabelText('The Shielded Site'));
  expect(queryByLabelText('Loading')).toBeNull();
});

test('clicking close button calls onRequestClose callback', () => {
  const close = jest.fn();
  const { getByLabelText } = render(<Modal onRequestClose={close} />);
  userEvent.click(getByLabelText('Close'));
  expect(close).toHaveBeenCalledTimes(1);
});

test('pressing Escape key calls onRequestClose callback', () => {
  const close = jest.fn();
  const { getByRole } = render(<Modal onRequestClose={close} />);
  userEvent.type(getByRole('dialog'), '{esc}');
  expect(close).toHaveBeenCalledTimes(1);
});
