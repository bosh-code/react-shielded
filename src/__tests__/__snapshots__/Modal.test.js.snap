// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders and matches snapshot 1`] = `null`;
